src/ModelContextProtocol/Shared/McpSession.cs

```diff
@@ -152,23 +161,89 @@ await _transport.SendMessageAsync(new JsonRpcError
 
     private async Task HandleMessageAsync(IJsonRpcMessage message, CancellationToken cancellationToken)
     {
-        switch (message)
+        using Activity? activity = s_activitySource.StartActivity("HandlingMessage");
+        TagList tags = default;
+
+        Stopwatch? measuring = s_operationDurationHistogram.Enabled ? Stopwatch.StartNew() : null;
+        bool addTags = activity is not null || measuring is not null;
+
+        try
         {
-            case JsonRpcRequest request:
-                await HandleRequest(request, cancellationToken).ConfigureAwait(false);
-                break;
+            if (addTags)
+            {
+                tags.Add("mcp.session.id", _id);
+            }
+
+            switch (message)
+            {
+                case JsonRpcRequest request:
+                    if (addTags)
+                    {
+                        tags.Add("mcp.request.id", request.Id.ToString());
+                        tags.Add("mcp.request.method", request.Method);
+
+                        if (request.Params is JsonElement je)
+                        {
+                            switch (request.Method)
+                            {
+                                case RequestMethods.ToolsCall:
+                                case RequestMethods.PromptsGet:
```

Comment:
Can we update the Activity display name based on the request.method?

Reply:
The name is specified at Activity creation time. The description is what is shown in the Aspire Dashboard.